dll/win32/shell32/folders/CFSFolder.cpp

```diff
@@ -129,67 +129,117 @@ HRESULT GetCLSIDForFileType(PCUIDLIST_RELATIVE pidl, LPCWSTR KeyName, CLSID* pcl
+        PathAppendW(wszIniFullPath, wszDesktopIni);
 
-        PathAppendW(wszFolderPath, wszDesktopIni);
+        // get current folder
+        WCHAR wszCurDir[MAX_PATH];
+        GetCurrentDirectoryW(_countof(wszCurDir), wszCurDir);
 
-        if (PathFileExistsW(wszFolderPath))
+        WCHAR wszValue[MAX_PATH], wszTemp[MAX_PATH];
+        if (getShellClassInfo(iconFile, wszValue, _countof(wszValue), wszIniFullPath))
         {
-            WCHAR wszPath[MAX_PATH];
-            WCHAR wszCLSIDValue[CHARS_IN_GUID];
+            // wszValue --> wszTemp
+            ExpandEnvironmentStringsW(wszValue, wszTemp, _countof(wszTemp));
 
-            if (GetPrivateProfileStringW(shellClassInfo, iconFile, NULL, wszPath, MAX_PATH, wszFolderPath))
+            if (SetCurrentDirectoryW(wszFolderPath))
             {
-                ExpandEnvironmentStringsW(wszPath, szIconFile, cchMax);
+                // wszTemp --> szIconFile
+                GetFullPathNameW(wszTemp, cchMax, szIconFile, NULL);
+                SetCurrentDirectoryW(wszCurDir);
 
-                *piIndex = GetPrivateProfileIntW(shellClassInfo, iconIndex, 0, wszFolderPath);
-                return S_OK;
-            }
-            else if (GetPrivateProfileStringW(shellClassInfo, clsid, NULL, wszCLSIDValue, CHARS_IN_GUID, wszFolderPath) &&
-                HCR_GetIconW(wszCLSIDValue, szIconFile, NULL, cchMax, &icon_idx))
-            {
-                *piIndex = icon_idx;
+                *piIndex = GetPrivateProfileIntW(s_shellClassInfo, iconIndex, 0, wszIniFullPath);
                 return S_OK;
             }
-            else if (GetPrivateProfileStringW(shellClassInfo, clsid2, NULL, wszCLSIDValue, CHARS_IN_GUID, wszFolderPath) &&
-                HCR_GetIconW(wszCLSIDValue, szIconFile, NULL, cchMax, &icon_idx))
+        }
+        else if (getShellClassInfo(clsid, wszValue, _countof(wszValue), wszIniFullPath) &&
+                 HCR_GetIconW(wszValue, szIconFile, NULL, cchMax, piIndex))
+        {
+            return S_OK;
+        }
+        else if (getShellClassInfo(clsid2, wszValue, _countof(wszValue), wszIniFullPath) &&
+                 HCR_GetIconW(wszValue, szIconFile, NULL, cchMax, piIndex))
+        {
+            return S_OK;
+        }
+        else if (getShellClassInfo(iconResource, wszValue, _countof(wszValue), wszIniFullPath))
+        {
+            *piIndex = PathParseIconLocationW(wszValue);
+
+            // wszValue --> wszTemp
+            ExpandEnvironmentStringsW(wszValue, wszTemp, _countof(wszTemp));
+
+            if (SetCurrentDirectoryW(wszFolderPath))
             {
-                *piIndex = icon_idx;
+                // wszTemp --> szIconFile
+                GetFullPathNameW(wszTemp, cchMax, szIconFile, NULL);
+                SetCurrentDirectoryW(wszCurDir);
                 return S_OK;
             }
         }
-    }
-
-    static const WCHAR folder[] = { 'F', 'o', 'l', 'd', 'e', 'r', 0 };
+    } while (0);
```

Comment:
Please avoid that construct in main code. This is part of the few exceptions where the "break;" you have above can be replaced by a "goto Quit;", and a "Quit:" label be used before the "getDefaultIconLocation2 call.

Reply:
OK.